### PR TITLE
Feat/mempool tx timeout and rebroadcast for raft

### PR DESCRIPTION
**Pull request by Alexader**

### CODE REVIEW

File: pkg/order/mempool/btree_index.go [hunk 1]
```diff
@@ -22,6 +24,17 @@ func (oik *orderedIndexKey) Less(than btree.Item) bool {
 	return oik.nonce < other.nonce
 }
 
+type sortedTtlKey struct {
```
Comment (Lizen0512) [inline pkg/order/mempool/btree_index.go:27]: the struct is same as structure orderedTimeoutKey, can you merge them？

File: config/order.toml [hunk 2]
```diff
@@ -1,5 +1,6 @@
 [raft]
 batch_timeout               = "0.3s"  # Block packaging time period.
+rebroadcast_timeout         = "3m" # Node timeout tick to check if there exist txs to be rebroadcast
```
Comment (Lizen0512) [inline config/order.toml:3]: Please change it to ‘tolerance_time’ to be consistent with the rbft configuration naming.

File: pkg/order/mempool/mempool_impl.go [hunk 3]
```diff
@@ -248,5 +276,75 @@ func (mpi *mempoolImpl) processCommitTransactions(state *ChainState) {
 	}
 	mpi.logger.Debugf("Replica %d removes batches in mempool, and now there are %d non-batched txs,"+
 		"priority len: %d, parkingLot len: %d, batchedTx len: %d, txHashMap len: %d", mpi.localID, mpi.txStore.priorityNonBatchSize,
-		 mpi.txStore.priorityIndex.size(), mpi.txStore.parkingLotIndex.size(), len(mpi.txStore.batchedTxs), len(mpi.txStore.txHashMap))
+		mpi.txStore.priorityIndex.size(), mpi.txStore.parkingLotIndex.size(), len(mpi.txStore.batchedTxs), len(mpi.txStore.txHashMap))
+}
+
+func (mpi *mempoolImpl) GetTimeoutTransactions(rebroadcastDuration time.Duration) [][]*pb.Transaction {
+	// all the tx whose live time is less than lowBoundTime should be rebroadcast
+	mpi.logger.Debugf("------- Start gathering timeout txs, ttl index len is %d -----------", mpi.txStore.ttlIndex.index.Len())
```
Comment (Lizen0512) [inline pkg/order/mempool/mempool_impl.go:284]: Remove the special characters ‘------- ’

File: pkg/order/mempool/mempool_impl.go [hunk 4]
```diff
@@ -91,10 +101,17 @@ func (mpi *mempoolImpl) processDirtyAccount(dirtyAccounts map[string]bool) {
...
-			// inset ready txs into priorityIndex.
+			// insert ready txs into priorityIndex and set ttlIndex for these txs.
 			for _, tx := range readyTxs {
-				mpi.txStore.priorityIndex.insertByOrderedQueueKey(account, tx)
+				if !mpi.txStore.priorityIndex.data.Has(makeTimeoutKey(account, tx)) {
+					mpi.txStore.priorityIndex.insertByTimeoutKey(account, tx)
+					if list.items[tx.Nonce].local {
+						// no need to rebroadcast tx from other nodes to reduce network overhead
+						mpi.txStore.ttlIndex.insertByTtlKey(account, tx.Nonce, tx.Timestamp)
+					}
```
Comment (Lizen0512) [inline pkg/order/mempool/mempool_impl.go:111]: the insertByTtlKey() func shoud be called in method insertTxs, ttlIndex is not directly related to the ready queue.

File: pkg/order/mempool/mempool_impl.go [hunk 5]
```diff
@@ -248,5 +276,75 @@ func (mpi *mempoolImpl) processCommitTransactions(state *ChainState) {
...
+func (mpi *mempoolImpl) GetTimeoutTransactions(rebroadcastDuration time.Duration) [][]*pb.Transaction {
+	// all the tx whose live time is less than lowBoundTime should be rebroadcast
+	mpi.logger.Debugf("------- Start gathering timeout txs, ttl index len is %d -----------", mpi.txStore.ttlIndex.index.Len())
+	currentTime := time.Now().UnixNano()
+	if currentTime < mpi.txStore.earliestTimestamp+rebroadcastDuration.Nanoseconds() {
+		// if the latest incoming tx has not exceeded the timeout limit, then none will be timeout
+		return [][]*pb.Transaction{}
+	}
+
+	txList := make([]*pb.Transaction, 0)
```
Comment (Lizen0512) [inline pkg/order/mempool/mempool_impl.go:291]: It is seems that txlist is not needed, and shardTxList can be divided by timeoutItems.

File: pkg/order/mempool/mempool_impl.go [hunk 6]
```diff
@@ -248,5 +276,75 @@ func (mpi *mempoolImpl) processCommitTransactions(state *ChainState) {
...
+		txMap, ok := mpi.txStore.allTxs[item.account]
+		if !ok || currentTime < timeoutTime {
+			return false
+		}
+		tx := txMap.items[item.nonce].tx
+		txList = append(txList, tx)
+		timeoutItems = append(timeoutItems, item)
+		return true
+	})
+	if len(txList) == 0 {
```
Comment (Lizen0512) [inline pkg/order/mempool/mempool_impl.go:311]: The transaction list is definitely not empty.

File: pkg/order/mempool/mempool_impl.go [hunk 7]
```diff
@@ -248,5 +276,75 @@ func (mpi *mempoolImpl) processCommitTransactions(state *ChainState) {
...
+	if totalLen%batchLen != 0 {
+		batchNums++
+	}
+	shardedLists := make([][]*pb.Transaction, 0, batchNums)
+	for i := uint64(0); i <= batchNums; i++ {
+		actualLen := batchLen
+		if end-begin+1 < batchLen {
+			actualLen = end - begin + 1
+		}
+		if actualLen == 0 {
```
Comment (Lizen0512) [inline pkg/order/mempool/mempool_impl.go:339]: When will actualLen be 0？

**Review by Lizen0512 — APPROVED** [review]

File: pkg/order/mempool/tx_store.go [hunk 8]
```diff
@@ -171,3 +193,35 @@ func (nc *nonceCache) setPendingNonce(account string, nonce uint64) {
...
+
+func newTxLiveTimeMap() *txLiveTimeMap {
+	return &txLiveTimeMap{
+		index: btree.New(btreeDegree),
+		items: make(map[string]int64),
+	}
+}
+
+func (tlm *txLiveTimeMap) insertByTtlKey(account string, nonce uint64, liveTime int64) {
+	tlm.index.ReplaceOrInsert(&orderedTimeoutKey{account, nonce, liveTime})
```
Comment (Lizen0512) [inline pkg/order/mempool/tx_store.go:212]: The transaction with previous timestamp should be deleted.

**Review by Lizen0512 — APPROVED** [review]

File: pkg/order/mempool/mempool_impl.go [hunk 9]
```diff
@@ -131,18 +146,22 @@ func (mpi *mempoolImpl) generateBlock() (*raftproto.RequestBatch, error) {
 		if txSeq >= 1 {
 			_, seenPrevious = mpi.txStore.batchedTxs[orderedIndexKey{account: tx.account, nonce: txSeq - 1}]
 		}
+		if txSeq == 3 {
+			mpi.logger.Infof("seenPrevious %v and commitNonce is %d", seenPrevious, commitNonce)
+			mpi.logger.Infof("batched txs is %v", mpi.txStore.batchedTxs)
+		}
```
Comment (jzhe886) [inline pkg/order/mempool/mempool_impl.go:152]: why print the log here?

File: pkg/order/mempool/mempool_impl.go [hunk 10]
```diff
@@ -248,5 +272,67 @@ func (mpi *mempoolImpl) processCommitTransactions(state *ChainState) {
...
+	}
+
+	timeoutItems := make([]*orderedTimeoutKey, 0)
+	mpi.txStore.ttlIndex.index.Ascend(func(i btree.Item) bool {
+		item := i.(*orderedTimeoutKey)
+		if item.timestamp > math.MaxInt64 {
+			// TODO(tyx): if this tx has rebroadcast many times and exceeded a final limit,
+			// it is expired and will be removed from mempool
+			return true
+		}
```
Comment (jzhe886) [inline pkg/order/mempool/mempool_impl.go:294]: item.timestamp not gather than math.MaxInt64

**Review by jzhe886 — APPROVED** [review]

**Pull request merged**